### PR TITLE
Bugfix/up memory usage good enough

### PR DESCRIPTION
**Pull request by pimotte**
While the name of this branch is not very enthousiastic, this pull request makes it very hard to crash due to high memory usage during an UpOperation. The problem used to be the huge amounts of metadata cached in memory, so all of these caches have been transformed into database-queries.

Upping 300K files of 512 bytes each to a local repo takes the following times when limiting RAM to 512MB.
real    47m0.783s
user    10m39.213s
sys 1m54.387s

The majority of this time is the querying of the database during indexing and the committing of the write operations to the database.

### CODE REVIEW

File: syncany-lib/src/main/java/org/syncany/database/dao/FileHistorySqlDao.java [hunk 1]
```diff
@@ -166,6 +168,111 @@ public FileHistoryId expandFileHistoryId(FileHistoryId fileHistoryIdPrefix) {
...
+		try (PreparedStatement preparedStatement = getStatement("filehistory.select.master.findLatestFileVersionsForPath.sql")) {
+			preparedStatement.setString(1, path);
+			try (ResultSet resultSet = preparedStatement.executeQuery()) {
+				// Fetch the latest versions of all files that once existed with the given path and find the most recent by comparing vector clocks
+				String latestFileHistoryId = null;
+				Long latestFileVersion = null;
+				VectorClock latestVectorClock = null;
+				while (resultSet.next()) {
+					VectorClock clock = VectorClock.parseVectorClock(resultSet.getString("vectorclock_serialized"));
+					if (latestVectorClock == null || VectorClock.compare(clock, latestVectorClock) == VectorClock.VectorClockComparison.GREATER) {
```
Comment (binwiederhier) [inline syncany-lib/src/main/java/org/syncany/database/dao/FileHistorySqlDao.java]: Very elegant. I never thought of that :+1:

File: syncany-lib/src/main/java/org/syncany/database/dao/FileHistorySqlDao.java [hunk 2]
```diff
@@ -166,6 +168,111 @@ public FileHistoryId expandFileHistoryId(FileHistoryId fileHistoryIdPrefix) {
...
+					}
+				}
+
+				// If no active file history exists for this path, return
+				if (latestFileHistoryId == null) {
+					return null;
+				}
+
+				// Find for the latest file history the last file version to check if the file has moved
+				PartialFileHistory fileHistory = getLastVersionByFileHistoryId(latestFileHistoryId);
```
Comment (binwiederhier) [inline syncany-lib/src/main/java/org/syncany/database/dao/FileHistorySqlDao.java]: I don't understand this part at all. Can you elaborate on this?

File: syncany-lib/src/main/java/org/syncany/operations/up/Indexer.java [hunk 3]
```diff
@@ -91,10 +89,18 @@
 
 	private LocalEventBus eventBus;
 
+
 	public Indexer(Config config, Deduper deduper) {
 		this.config = config;
 		this.deduper = deduper;
 		this.localDatabase = new SqlDatabase(config);
+		try {
+			localDatabase.getConnection().setReadOnly(true);
```
Comment (binwiederhier) [inline syncany-lib/src/main/java/org/syncany/operations/up/Indexer.java]: try/catch block in a constructor with `e.printStackTrace()`? Shouldn't this be something like `SqlDatabase.createReadOnly(config);` or something?

File: syncany-lib/src/main/java/org/syncany/operations/up/Indexer.java [hunk 4]
```diff
@@ -367,11 +330,21 @@ private void addFileVersion(FileProperties fileProperties) {
 			PartialFileHistory lastFileHistory = guessLastFileHistory(fileProperties);
 			FileVersion lastFileVersion = (lastFileHistory != null) ? lastFileHistory.getLastVersion() : null;
 
-			// 2. Create new file history/version
+			// 2. If file type changed, "close" the old file history by adding a file version that deletes the old file/directory
+			PartialFileHistory deletedFileHistory = null;
```
Comment (binwiederhier) [inline syncany-lib/src/main/java/org/syncany/operations/up/Indexer.java]: This seems to make sense, but it is not related to the memory issues, is it? Did you discover an issue here?

File: syncany-lib/src/main/resources/org/syncany/database/sql/filehistory.select.master.findLatestFileVersionsForPath.sql [hunk 5]
```diff
@@ -0,0 +1,18 @@
+SELECT fvmax.version AS version,
```
Comment (binwiederhier) [inline syncany-lib/src/main/resources/org/syncany/database/sql/filehistory.select.master.findLatestFileVersionsForPath.sql]: Very smart. A bit of documentation would be nice for this (I've added a bit)

**Comment by binwiederhier** [comment]
Do we have comparisons of these times with the old code? When I tested it back in the day, 200k wasn't even possible, but that was before the batchification.

Apart from the comments I added, it looks good to me. I've changed a few things -- all minor and mostly style-related: https://github.com/syncany/syncany/commit/a7243d326eb6779fa0ee196a2de25dcd8502acc8

It's a bit light on the tests and some of the code is not very pretty, but I think we need to do some efforts to tackle that anyway, I think.

File: syncany-lib/src/main/java/org/syncany/database/dao/FileHistorySqlDao.java (hunk 2 above)
Comment (binwiederhier) [inline syncany-lib/src/main/java/org/syncany/database/dao/FileHistorySqlDao.java]: By this part I mean this:

```java
// Find for the latest file history the last file version to check if the file has moved
PartialFileHistory fileHistory = getLastVersionByFileHistoryId(latestFileHistoryId);
				
if (fileHistory.getLastVersion().getVersion() == latestFileVersion) {
	return fileHistory;
}
else {
	return null;
}
```

**Comment by pimotte** [comment]
These are not comparisons, but different methods of time measurement. Ie. real world time and computational time.

**Comment by binwiederhier** [comment]
I know that these are not comparisons, it's an output of `time`. That's why I was asking for comparisons :smile:

**Comment by pimotte** [comment]
Yes. Reading. Post-batchificiation crashes at about 150K files due to lack of memory, so the comparison is against Infinity/Infinity/Infinity :D

File: syncany-lib/src/main/java/org/syncany/operations/up/Indexer.java (hunk 3 above)
Comment (pimotte) [inline syncany-lib/src/main/java/org/syncany/operations/up/Indexer.java]: Agreed, I'll fix this.

File: syncany-lib/src/main/java/org/syncany/operations/up/Indexer.java (hunk 4 above)
Comment (pimotte) [inline syncany-lib/src/main/java/org/syncany/operations/up/Indexer.java]: Previously we would do the file deletion by iterating over the entire database (https://github.com/syncany/syncany/blob/develop/syncany-lib/src/main/java/org/syncany/operations/up/Indexer.java#L210). I needed to remove the cache that was used here, and I switched to using the output from Status to do deletion. However, the StatusOperation does not detect type changes and having a file twice in the changeSet (one time as added, one time as deleted) seemed a bit dangerous.

File: syncany-lib/src/main/java/org/syncany/database/dao/FileHistorySqlDao.java (hunk 2 above)
Comment (pimotte) [inline syncany-lib/src/main/java/org/syncany/database/dao/FileHistorySqlDao.java]: That comment made way more sense in context with how I thought about it with Tim. I've adjusted it to make sense out of context. Is it understandable now?

**Comment by binwiederhier** [comment]
Without trying to be too picky, boolean flags are not so nice: `createConnection(false)` does isn't very clear. Can we change that to `createReadOnlyConnection()` and `createConnection()` or something like that?

**Comment by binwiederhier** [comment]
Once that is done, feel free to merge this! Good stuff.

**Comment by jdonkervliet** [comment]
@pimotte if you merge this, I will continue #451.
I want to be able to check if my changes break any tests, and #451 is currently expecting some changes from this branch.

**Comment by pimotte** [comment]
Great, I'll get to this on monday/tuesday

**Comment by pimotte** [comment]
@binwiederhier, I've been thinking about the readonly thing a little bit (and this is bikeshed to me, so if you disagree I'll do it the way you suggested it).

Having two seperate methods might be clearer if you just read the plain code, but if you include the parameter signature it's just as clear what it does. Removing the flag will lead to a little bit of code duplication, but even moreso if at some point we want to be able to have SQLDatabases with, say, different transaction isolation level.

Hence, I feel like it would be cleaner to add a little bit of javadoc and leave the functions as they are, so this is what I'd like to do, but feel free to overrule me.

**Comment by pimotte** [comment]
Discussion on IRC has lead to the conclusion that the boolean-flag situation is not ideal, but not important enough to hold this up. If anyone reading this has ideas for a clear, clean-code solution, feel free to contact us/PR.

Merging this:D

**Comment by jdonkervliet** [comment]
I will finish #451 *soon*.

Also, I have been looking through old pull requests for the software architecture course, but I'm not so sure this happens frequently in general. If you want the boolean flag issue to be resolved at some point, it might be worth it to create an issue or insert a TODO in the code?

**Pull request merged**